Show poll-failure details

PIG208 opened the pull request:
Ready for review.

This fixes #555.

PIG208 commented:
A question to be answered: how do we determine if an error is transient?

Currently we handle 3 classes of errors in `UpdateMachine`:
1. `ZulipApiException`: The event queue is lost. This is commonly caused by inactivity.
1. `Server5xxException`, `NetworkException`: A server-side bug, or hiccups due to unstable Internet connection.
1. default: Could be anything that happens in the client code during the `getEvents` call.

I think it is usually safe to assume that 1 is recoverable, while we are less certain about that for 2 and 3. An idea is to define a number of retries for the client to start displaying error messages. And then in a follow-up PR the client can try to do something different (which is #186).

PIG208 commented:
For non-transient errors, we will probably show a modal instead.

gnprice commented:
If there's a `ZulipApiException` *other* than the event queue being gone, then that's likely non-transient — it's either a bug in the client, or some kind of unexpected incompatibility between server and client. A retry is likely to just hit the same error again.

(Those fall into the `default` case.)

gnprice commented:
Other than that point, the description above all looks right.

gnprice commented on lib/model/store.dart:
```diff
@@ -777,22 +790,27 @@ class UpdateMachine {
-            // TODO tell user if transient polling errors persist
             // TODO reset to short backoff eventually
+            accumlatedTransientFailureCount++;
+            if (accumlatedTransientFailureCount > transientFailureCountNotifyThreshold) {
+              reportErrorToUserInDialog('Failed to reach server. Will retry: $e');
```
Let's leave this case as a toast. It's a server 5xx or a network exception, so it's unlikely there's anything the client should be doing differently about it — no need to get in the user's way and potentially annoy them.

gnprice replied:
Then let's start showing the toasts once it's been a few seconds since we started retrying. The simple way to do that is to make the count threshold about 5, rather than 10 — at the 6th failure, it's been an expected 1.55s since the first failure, and we'll be backing off for an expected 1.6s before the next retry.

gnprice reviewed (commented):
(Quick comments on the high-level behavior.)

chrisbobbe commented on lib/log.dart:
```diff
@@ -30,3 +33,45 @@ bool debugLog(String message) {
   return true;
 }
+
+/// Display an error message.
```
How about "Display an error message in a [SnackBar]." I like how the summary line on `reportErrorToUserInDialog`'s dartdoc says how the error is displayed.

chrisbobbe commented on lib/log.dart:
```diff
@@ -30,3 +33,45 @@ bool debugLog(String message) {
+
+/// Display an error message.
+///
+/// This shows a [SnackBar] containing the message after app startup,
+/// otherwise logs it to the console.
```
Can we be more specific about "after app startup"? I can imagine confused readers wondering how the snackbar could be shown at any time other than after the app starts up.

chrisbobbe commented on lib/log.dart:
```diff
@@ -30,3 +33,45 @@ bool debugLog(String message) {
+/// This shows a [SnackBar] containing the message after app startup,
+/// otherwise logs it to the console.
+///
+/// See also: [ZulipApp._reportErrorToUserBriefly]
```
This is a private method, and also kind of feels like an implementation detail. Is this the kind of information that callers need, to understand if `reportErrorToUserBriefly` is right for them, and what its expectations/guarantees are like? (Similarly with the dartdoc on `reportErrorToUserInDialog`.)

PIG208 replied:
The primary goal of including this line is to have a back link to where the setters get used, since it is not obvious. I think it would be more appropriate to exclude that bit from the dart doct, though, considering the point that it's some implementation detail.

chrisbobbe commented on lib/log.dart:
```diff
@@ -30,3 +33,45 @@ bool debugLog(String message) {
+/// otherwise logs it to the console.
+///
+/// See also: [ZulipApp._reportErrorToUserBriefly]
+void Function(String message) get reportErrorToUserBriefly => (message) {
```
How about `reportErrorToUserInSnackbar`, to be more informative, and more parallel with `reportErrorToUserInDialog`?

PIG208 replied:
Yeah that sounds like a good choice.

chrisbobbe commented on lib/log.dart:
```diff
@@ -30,3 +33,45 @@ bool debugLog(String message) {
+
+String? get debugLastReportedError => _debugLastReportedError;
+String? _debugLastReportedError;
+String? takeLastReportedError() {
```
Should this be `debugTakeLastReportedError`?

chrisbobbe commented on test/widgets/app_test.dart:
```diff
@@ -166,5 +167,49 @@ void main() {
+      // After app startup, reportErrorToUserInDialog displays a dialog.
+      reportErrorToUserInDialog('test error message');
+      await tester.pump();
+      check(finder.evaluate()).single;
```
Could maybe be strengthened with `checkErrorDialog`; just a thought.

PIG208 replied:
Could be a bit repetitive, but it should be good to also check the title.

chrisbobbe commented on test/model/store_test.dart:
```diff
@@ -427,53 +427,53 @@ void main() {
-    test('retries on Server5xxException', () {
-      checkRetry(() => connection.prepare(httpStatus: 500, body: 'splat'));
-    });
+    group('retries', () {
```
```
store test [nfc]: Make a group for retry tests.
```

Let's leave off the `[nfc]` in the commit message: https://github.com/zulip/zulip-flutter/pull/870#discussion_r1717639672

chrisbobbe commented on lib/model/store.dart:
```diff
@@ -801,6 +801,7 @@ class UpdateMachine {
             assert(debugLog('Error polling event queue for $store: $e\n'
                 'Backing off and retrying even though may be hopeless…'));
             // TODO tell user on non-transient error in polling
+            reportErrorToUserInDialog('Error loading server data. Will retry: $e');
```
```
store: Reoprt non-transient polling errors.
```

commit-message nit: spelling

chrisbobbe replied:
Also this TODO can be removed:

```dart
            // TODO tell user on non-transient error in polling
```

chrisbobbe replied:
Also, can this be a translated string? If that's not straightforward, let's include a `TODO(i18n)`.

chrisbobbe replied:
Hmm. And for non-transient errors, we expect that retrying won't help, right? It seems like we should have a TODO to not retry. I'm also wondering if retrying _and_ showing a dialog, for a task that's likely to be hopeless, will spam the user with the same dialog repetitively.

PIG208 replied:
> Hmm. And for non-transient errors, we expect that retrying won't help, right? It seems like we should have a TODO to not retry. I'm also wondering if retrying and showing a dialog, for a task that's likely to be hopeless, will spam the user with the same dialog repetitively.

Yes, that's #186. Added a line mentioning that.

chrisbobbe reviewed (commented):
Thanks! Comments below.

Then one other comment is, for all the new poll-failure feedback we're giving the user, let's identify which server it's about. This is particularly relevant when the issue is a network problem, but applies whenever the feedback can be significantly delayed. The user might have given up waiting for this account, like a minute or more ago, and is doing other things—maybe even in a different account—when the feedback appears.

PIG208 commented:
The PR has been updated. Thanks for the review!

I'm not entirely sure about the quality of the translation strings. These messages will appear more technical at the current stage for the ease of debugging. However, polishing user facing errors might be a post-launch task.

PIG208 commented:
Updated the PR to avoid wrapping `test` calls in `checkRetry`.

gnprice commented on lib/log.dart:
```diff
@@ -1,3 +1,6 @@
+import 'package:flutter/material.dart';
+
+import 'widgets/app.dart';
```
This `lib/log.dart` file is one that virtually the whole app imports, including parts of `lib/api/`. So I'd like to keep it from depending on the app's widgets.

That's the motivation for the indirection found in my draft branch that IIRC you started from:
https://github.com/zulip/zulip-flutter/compare/main...gnprice:zulip-flutter:dev-retry#diff-055b57342eabea1b640e04ca24eb50bc75d0c61383b928ad7211e37159a6c57bR34-R42
https://github.com/zulip/zulip-flutter/commit/68682403587804ab6dd392491fd7700596972ea3#diff-4c278608930a95a1cba05bce5a0f7d8f43bae364ed1e722eca5d8c1c4b29b57aR94-R101

where this file just has a mutable callback that defaults to something simple, and then ZulipApp can set the callback to something more complex after it starts up.

(In fact this file doesn't end up using these imports, right? It still has that indirection. I suspect the reason the linter hasn't complained about them is just the references in dartdoc.)

PIG208 replied:
> In fact this file doesn't end up using these imports, right?

Right, these imports were included only for the dart doc references. Removing them and updating the commit message to reflect the motivation:

```diff
-    The imports in `lib/log.dart` are for the code references in
-    the reportErrorToUser* helpers' dart doc.
-
-    The motivation of having these set of helpers, rather than using
-    showErrorDialog and showSnackBar directly, is that we make displaying
-    errors more accessible in non-widget code.
+    The motivation of having this indirection, rather than using
+    `showErrorDialog` and `showSnackBar` directly, is to keep dependencies
+    of `lib/log.dart` from relying on widget code.    
```

PIG208 commented:
Updated the PR to adjust comments on `lib/log.dart` and made dialog title a parameter. I also tweaked the formatting and made some other minor improvements.

PIG208 commented:
This update modifies the translation strings to include the realm URL in error messages.

gnprice commented:
That fix LGTM on a skim, thanks. I'll leave @chrisbobbe to do the next round of full review.

chrisbobbe commented on lib/log.dart:
```diff
@@ -30,3 +30,48 @@ bool debugLog(String message) {
+
+/// Display an error message in a [SnackBar].
+///
+/// This shows a [SnackBar] containing the message after [ZulipApp] becomes ready,
+/// otherwise logs it to the console.
```
```suggestion
/// This shows a [SnackBar] containing the message if [ZulipApp] is ready,
/// otherwise logs it to the console.
```

(similarly with `reportErrorToUserInDialog` below)

chrisbobbe commented on lib/log.dart:
```diff
@@ -30,3 +30,48 @@ bool debugLog(String message) {
+///
+/// This shows a [SnackBar] containing the message after [ZulipApp] becomes ready,
+/// otherwise logs it to the console.
+// This and `reportErrorToUserInDiaglog` get set in [ZulipApp].  We need this
```
```suggestion
// This and `reportErrorToUserInDialog` get set in [ZulipApp].  We need this
```

chrisbobbe commented on assets/l10n/app_en.arb:
```diff
@@ -160,6 +160,31 @@
+      "error": {"type": "String", "example": "Network request failed: HTTP status 500"}
+    }
+  },
+  "errorReconnectingToServer": "Reconnecting to server at {url}...",
```
How about "Reconnecting to {url}…"? It's a bit shorter, and also I think we prefer "…" over "...".

PIG208 replied:
Yeah, we can do a tiny sweep to replace "..." with "…"  in our translation strings. I saw the character used in tests and comments, but not sure why we didn't do that in `app_en.arb`.

PIG208 replied:
Opened #911.

chrisbobbe commented on lib/widgets/app.dart:
```diff
@@ -63,22 +65,52 @@ class ZulipApp extends StatefulWidget {
+
+  static void _reportErrorToUserInDialog(String title, String message) {
+    final context = navigatorKey.currentContext;
+    if (context == null) return;
```
I wonder when this might happen, and whether we want to log to the console in this case, in case there's a developer watching. (Like we do for the case where `ZulipApp` isn't ready.)

PIG208 replied:
Hmm, good question. The guarantee we enforce is that when `_ready` is `true`, `currentContext` is always non-null (see `ZulipApp._declareReady`), so it would be a bug if `context` is `null` here. We should use a `!` here and in the other method instead. I also added an assertion `assert(_ready.value)` just to make the invariant explicit.

chrisbobbe commented on assets/l10n/app_en.arb:
```diff
@@ -160,6 +160,31 @@
+  },
+  "errorConnectingToServer": "Failed to reach server at {url}. Will retry:\n\n{error}",
+  "@errorConnectingToServer": {
+    "description": "Dialog message when the the server cannot be reached",
```
```suggestion
    "description": "Dialog message when the server cannot be reached",
```

chrisbobbe reviewed (commented):
Thanks! Comments below, all small.

PIG208 commented:
Thanks for the comments. This should be ready for review again!

gnprice commented:
@PIG208 @chrisbobbe I recommend installing this PR version on your main devices for a while, with `flutter run --release`. I just installed it yesterday, and the output is interesting but can definitely use some tweaking to be less noisy (particularly with dialogs).

PIG208 commented:
Regarding the noisiness of the error dialog, I think a better UX is to present a snackbar with a brief error summary and a "Show more" button that brings up the dialog page with the full error message.

PIG208 commented:
Implemented the proposed change in a squash! commit. Before moving on to the next review, I would like to gather some feedback on this new UX ([discussion](https://chat.zulip.org/#narrow/stream/48-mobile/topic/error.20reporting/near/1931665))

chrisbobbe commented on lib/model/store.dart:
```diff
@@ -786,9 +789,12 @@ class UpdateMachine {
         switch (e) {
           case ZulipApiException(code: 'BAD_EVENT_QUEUE_ID'):
             assert(debugLog('Lost event queue for $store.  Replacing…'));
+            reportErrorToUser(localizations.errorReconnectingToServer(serverUrl));
             await store._globalStore._reloadPerAccount(store.accountId);
             dispose();
```
How about hiding the snackbar as soon as we have a new event queue?

```suggestion
            reportErrorToUser(localizations.errorReconnectingToServer(serverUrl));
            await store._globalStore._reloadPerAccount(store.accountId);
            ZulipApp.scaffoldMessenger?.hideCurrentSnackBar();
            dispose();
```

(Or maybe `reportErrorToUser` can return a callback, or return the `ScaffoldFeatureController` returned by `ScaffoldMessengerState.showSnackBar`, that we can use to hide the snackbar if one was shown.)

In my manual testing, the reconnection to CZO is finished before the snackbar disappears, but the snackbar is still present and holding my attention for some time, making it seem (I think) like the reconnection is taking longer than it actually takes. Also, it looks like maybe `ScaffoldMessengerState.showSnackBar` lets us make the exit animation duration shorter than the entrance animation, with its `snackBarAnimationStyle` param, and that might also be something we could do to help make the app feel snappy and responsive.

PIG208 replied:
These all sound like great ideas, thanks.

chrisbobbe reviewed (commented):
Thanks! Small comments below, and I'm continuing to test this out manually on my device.

chrisbobbe commented:
Thanks! This revision LGTM. One thought is that it would be nice if the snackbar API gave us a way to identify a specific snackbar to remove, so we could be sure we're removing the one we intend to. But it looks like our choices are the "current" snackbar and all snackbars. The "current" snackbar I guess means the latest one from the queue, if it hasn't expired, and that could be right or wrong depending on the durations of the snackbars and how much time has passed.

PIG208 commented:
Yeah, I figured that tracking if the last snack bar has expired would reduce collateral dismissal of unrelared snackbars in most cases.

PIG208 commented:
Updated the PR to add a comment on the https://github.com/zulip/zulip-flutter/pull/868#discussion_r1733560511 implementation.

PIG208 commented:
Pushed to rebase.

chrisbobbe commented:
Thanks, LGTM! Over to Greg for his review.

PIG208 commented:
Added a commit removing a stale TODO, and squashed the two of the error reporting commits (for transient and non-transient errors) together.

gnprice commented:
> The public message history from CZO can be downloaded from [this protected dropbox link](https://www.dropbox.com/scl/fi/5ugemicgz6lwrsxscqvwk/corpuses.zip?rlkey=xdje3piu5scypt3glc0n4lbrz&st=muzlmp6t&dl=0). We probably will think of a better way of reusing this in the future.

(Was this comment intended for #917?)

gnprice commented on lib/log.dart:
```diff
@@ -30,3 +32,51 @@ bool debugLog(String message) {
+  (_reportErrorToUser ??= _defaultReportErrorToUser)(message, details: details);
+};
+ReportErrorCallback? _reportErrorToUser;
+set reportErrorToUser(ReportErrorCallback? value) =>
+  _reportErrorToUser = value;
```
nit: use block body

Generally a setter always gets a block body.

See https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#consider-using--for-short-functions-and-methods

gnprice commented on lib/log.dart:
```diff
@@ -30,3 +32,51 @@ bool debugLog(String message) {
+/// A record containing the error message and error details from the last
+/// [reportErrorToUser] call.
+@visibleForTesting
+String? get debugLastReportedError => _debugLastReportedError;
+String? _debugLastReportedError;
+@visibleForTesting
+String? debugTakeLastReportedError() {
```
Just the "take" version should be enough.

gnprice commented on lib/log.dart:
```diff
@@ -30,3 +32,51 @@ bool debugLog(String message) {
+// This gets set in [ZulipApp].  We need this indirection to keep `lib/log.dart`
+// from importing widget code, because the file is a dependency for the rest of
+// the app.
+ReportErrorCallback get reportErrorToUser => (message, {details}) {
+  assert(() {
+    _debugLastReportedError = details != null ? '$message\n$details' : message;
+    return true;
+  }());
+  (_reportErrorToUser ??= _defaultReportErrorToUser)(message, details: details);
+};
```
This structure with a getter, that returns a fixed function expression, that mostly delegates to a function-typed field, that gets lazily initialized to a default value, feels a bit hard to understand.

It seems like the reason the getter is here, vs. just a field, is to inject this recording of `_debugLastReportedError`, to support `debugTakeLastReportedError`. Can we have the tests just look for the snack bar instead?

That'd make the tests more end-to-end, testing the actual user-facing behavior, which is good:
https://zulip.readthedocs.io/en/latest/testing/philosophy.html#integration-testing-or-unit-testing

And it seems like it'd simplify this code, too.

PIG208 replied:
It will indeed be a helpful refactor to simplify the entire thing to a single line.

This is mostly used for `test/model/store_test.dart` only, and we test it end-to-end in `test/widgets/app_test.dart`. I think we don't instantiate `TestZulipApp` for tests in `test/model`, but this can be easier resolved by assigning a custom `reportErrorToUserBriefly` that updates a variable like `checkNotified` does.

gnprice commented on assets/l10n/app_en.arb:
```diff
@@ -160,6 +160,41 @@
       "message": {"type": "String", "example": "Invalid format"}
     }
   },
+  "errorLoadingServerDataShort": "Error loading server data from {url}",
```
Let's try this wording:
```suggestion
  "errorLoadingServerDataShort": "Error connecting to Zulip. Retrying…",
```

and for the details:

`Error connecting to Zulip at {serverUrl}. Will retry:\n\n{error}`

Then use that for both the `Server5xxException() || NetworkException()` case and the default case. Both cases are about "an error loading server data", or "an error connecting to Zulip", anyway.

gnprice commented on test/model/store_test.dart:
```diff
@@ -427,53 +431,80 @@ void main() {
+    group('retries on errors', () {
+      void checkRetry(void Function() prepareError, {
+        required String errorMessage,
+        int expectedFailureCountNofityThreshold = 0,
```
```suggestion
        int expectedFailureCountNotifyThreshold = 0,
```

gnprice reviewed (commented):
Thanks @PIG208, and thanks @chrisbobbe for the previous reviews!

Some comments below. I've also just installed this PR branch on my main device, so I can see how this version of the UX works out over time.

PIG208 commented on lib/log.dart:
```diff
@@ -30,3 +32,51 @@ bool debugLog(String message) {
+// This gets set in [ZulipApp].  We need this indirection to keep `lib/log.dart`
+// from importing widget code, because the file is a dependency for the rest of
+// the app.
+ReportErrorCallback get reportErrorToUser => (message, {details}) {
```
Yeah, that makes sense. I previously created two functions `reportErrorToUserBriefly` (later renamed to `reportErrorToUserInSnackbar`), `reportErrorToUserInDialog`.

In this revision, these two functions were merged into one, losing the suffix, but technically, since it displays a snackbar either way, and only shows a dialog if you click the button on the snackbar, it should be still considered "brief".

PIG208 commented:
Thanks for the review! The PR has been updated.

gnprice commented on assets/l10n/app_en.arb:
```diff
@@ -160,6 +160,23 @@
       "message": {"type": "String", "example": "Invalid format"}
     }
   },
+  "errorLoadingServerDataShort": "Error connecting to Zulip. Retrying…",
```
```suggestion
  "errorConnectingToServerShort": "Error connecting to Zulip. Retrying…",
```

(didn't update that part in my previous suggestion, oops)

gnprice commented on test/model/store_test.dart:
```diff
@@ -391,6 +392,22 @@ void main() {
+      if (message == null) return;
+      lastReportedError = '$message\n$details';
+    }
+    // This overrides the function globally, but we don't need to worry about
+    // leaking the state because that is only relevant to widget tests.
+    // See `lib/log.dart ` for details.
+    reportErrorToUserBriefly = logAndReportErrorToUserBriefly;
```
Better to just systematically avoid leaking state between tests.

In general the way to do that is to (a) set it up from within a test case, and then (b) use `addTearDown` to ensure it gets reset when the test case is done. Given the function definition above, I think code like
```dart
  test('some test', () {
    reportErrorToUserBriefly = logAndReportErrorToUserBriefly;
    addTearDown(reportErrorToUserBriefly = defaultReportErrorToUserBriefly);
    // … rest of test …
```

should do it.

When the setup is more complicated than that, that's the sort of thing that we write helper functions named like "prepareFoo" for.

gnprice reviewed (commented):
Thanks!

The changes in the revision look good, modulo a couple of comments below.

I've also updated the build on my device, so I'll see how the UX works out.

Would you take the reportErrorToUserBriefly commit and the one it depends on, and send those as a separate PR (which this one can then stack on top of)? That way we can get those changes merged very soon, so they can be used in other features.

I'll take a pass over the commit messages but otherwise I think those two commits are basically ready. For the later commits that actually start giving the user feedback, I'll want a bit of bake time seeing the UX.

gnprice commented:
Quick observation on the UX:

I reopened the app after a while. It showed the loading indicator at top for a while, then successfully reconnected.

But then for some time after that, it showed a chain of several rounds of snack bars saying there was an error, followed by one saying "reconnecting". Each one lasts several seconds, and there were about three of them that started after the loading indicator finished.

I hit the "details" link on a couple of the errors. Both were NetworkException, saying the host name chat.zulip.org couldn't be resolved.

PIG208 commented:
Opened #935 for the two commits implementing the report error feature,
and #936 for the other nfc commit.

PIG208 commented:
Updated the PR to ignore `NetworkException` where the underlying cause is a `SocketException`, added some more documentation and cleaned up some diffs to make them easier to read. Feel free to install a build of this PR to test it out!

gnprice commented:
Thanks for the update! Installing on my phone now.

gnprice commented on test/model/store_test.dart:
```diff
@@ -455,53 +479,110 @@ void main() {
+            checkLastRequest(lastEventId: 1);
+            check(store).isLoading.isTrue();
+
+            if (i != numFailedRequests - 1) {
+              // Skip polling backoff unless this is the final iteration.
+              // This allows the next `updateMachine.debugAdvanceLoop` call to
+              // trigger the next request without wait.
+              async.flushTimers();
+            }
+
+            if (!expectNotifyError) {
+              // No matter how many retries there have been, no request should
+              // result in an error message.
+              check(takeLastReportedError()).isNull();
+              continue;
+            }
+            if (i < numFailedRequests - 1) {
+              // The error message should not appear until the `updateMachine`
+              // has retried the given number of times.
+              check(takeLastReportedError()).isNull();
+              continue;
+            }
```
The logic in this test (or set of tests) gets fairly tangly to follow.

I think the main thing is that it'd be helpful to test two aspects separately: there's the existing test cases that just check that we retry, and let's have separate new test cases that just check the error-reporting.

Then I think both sets of tests can be a lot simpler than this — most of the logic is only relevant for one aspect or the other. It's OK that some other fragments of code will be duplicated between them.

gnprice replied:
Separately and more tactically, I think it'd help to pull the last iteration here outside of the loop — have the loop only concern itself with the first N-1 requests. That would avoid these multiple conditionals about whether it's the last iteration.

gnprice commented on lib/model/store.dart:
```diff
@@ -811,26 +829,57 @@ class UpdateMachine {
             assert(debugLog('Transient error polling event queue for $store: $e\n'
                 'Backing off, then will retry…'));
-            // TODO tell user if transient polling errors persist
+            if (e.cause is! SocketException) {
+              // Heuristic check to only report interesting errors to the user.
+              // This currently ignores [SocketException], which typically
+              // occurs when the client goes back from sleep.
+              // TODO: Investigate if there are other cases of [SocketException]
+              //   that might turn out to be interesting.
+              //
+              // See also:
+              //  * [NetworkException.cause], which is the underlying exception.
```
```suggestion
            if (e.cause is! SocketException) {
              // Heuristic check to only report interesting errors to the user.
              // A [SocketException] is common when the app returns from sleep.
```

I think the TODO isn't actionable.

The "See also" isn't needed — `.cause` is just above, and the reader can follow that to its definition using an IDE.

Then while at it I tightened the sentence about SocketException. No need to say what the code does, as the code here speaks for itself; only to explain why we chose to make the code do that.

gnprice reviewed (commented):
OK, I've observed this on my phone for a while; and it's been pretty quiet, which is good.

After seeing it in action, I think we can leave out the last commit, with the "Reconnecting to {serverUrl}…" message. It feels redundant with the loading indicator at the top; and the fact that the toast disappears on its own schedule while the loading indicator can be still going seems incongruous, and likely to be confusing.

Otherwise this generally all looks good; just a few comments below.

(I also feel like the `UpdateMachine.poll` method has gotten kind of unwieldy; maybe I'll take a stab at reorganizing that after merging this PR. It's already kind of unwieldy in main.)

PIG208 commented:
Thanks for the feedback! Dropped the last commit and previous changes relevant to the tests. A new group of tests is now responsible for testing the error reporting feature.

gnprice commented on test/model/store_test.dart:
```diff
@@ -652,6 +655,90 @@ void main() {
       checkRetry(() => connection.prepare(httpStatus: 200, body: 'nonsense'));
     });
+
+    group('reportErrorToUserBriefly', () {
```
nit:
```suggestion
    group('error reporting', () {
```

Otherwise it sounds like these are tests of `reportErrorToUserBriefly`.

gnprice commented on test/model/store_test.dart:
```diff
@@ -652,6 +655,90 @@ void main() {
+        updateMachine.poll();
+      }
+
+      late void Function() prepareError;
+
+      void pollAndFail(FakeAsync async) {
+        prepareError();
+        updateMachine.debugAdvanceLoop();
```
```suggestion
      void pollAndFail(FakeAsync async) {
        updateMachine.debugAdvanceLoop();
```

Instead the callers can say `connection.prepare(…)` (i.e., the body of what they would set `prepareError` to) just before each call to `pollAndFail`. That isn't really any longer — it makes a very slight duplication of the content of those `connection.prepare` calls — and it makes the control flow substantially more straightforward.

gnprice commented on test/model/store_test.dart:
```diff
@@ -652,6 +655,90 @@ void main() {
+        for (int i = 0; i < UpdateMachine.transientFailureCountNotifyThreshold; i++) {
+          pollAndFail(async);
+          check(takeLastReportedError()).isNull();
+          // This would skip the pending polling backoff.
+          async.flushTimers();
```
```suggestion
          // This skips the pending polling backoff.
          async.flushTimers();
```

I.e. this does skip that backoff, right?

Or am I missing a subtlety where it doesn't actually do so, but only *would* do so if something were different?

gnprice reviewed (commented):
Thanks for the revision! Just small comments below, all in the tests.

And as a note to myself, there's a followup I mentioned at https://github.com/zulip/zulip-flutter/pull/868#pullrequestreview-2376436250 I might do after merge.

PIG208 commented:
Thanks! Ready for review now.

gnprice commented:
Looks good — merging.
